presto-main/src/main/java/com/facebook/presto/operator/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,770 @@
+                    nullChannel,
+                    outputBuffer,
+                    serdeFactory,
+                    maxMemory);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final Function<Page, Page> pagePreprocessor;
+    private final PagePartitioner pagePartitioner;
+    private final LocalMemoryContext systemMemoryContext;
+    private boolean finished;
+
+    public OptimizedPartitionedOutputOperator(
+            OperatorContext operatorContext,
+            List<Type> sourceTypes,
+            Function<Page, Page> pagePreprocessor,
+            PartitionFunction partitionFunction,
+            List<Integer> partitionChannels,
+            List<Optional<ConstantExpression>> partitionConstants,
+            boolean replicatesAnyRow,
+            OptionalInt nullChannel,
+            OutputBuffer outputBuffer,
+            PagesSerdeFactory serdeFactory,
+            DataSize maxMemory)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.pagePreprocessor = requireNonNull(pagePreprocessor, "pagePreprocessor is null");
+        this.pagePartitioner = new PagePartitioner(
+                partitionFunction,
+                partitionChannels,
+                partitionConstants,
+                replicatesAnyRow,
+                nullChannel,
+                outputBuffer,
+                serdeFactory,
+                sourceTypes,
+                maxMemory,
+                operatorContext.getDriverContext().getLifespan());
+
+        operatorContext.setInfoSupplier(this::getInfo);
+        this.systemMemoryContext = operatorContext.newLocalSystemMemoryContext(PartitionedOutputOperator.class.getSimpleName());
+        this.systemMemoryContext.setBytes(pagePartitioner.getInitialRetainedSizeInBytes());
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    public PartitionedOutputInfo getInfo()
+    {
+        return pagePartitioner.getInfo();
+    }
+
+    @Override
+    public void finish()
+    {
+        finished = true;
+        pagePartitioner.flush();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finished && isBlocked().isDone();
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        ListenableFuture<?> blocked = pagePartitioner.isFull();
+        return blocked.isDone() ? NOT_BLOCKED : blocked;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished && isBlocked().isDone();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        if (page.getPositionCount() == 0) {
+            return;
+        }
+
+        page = pagePreprocessor.apply(page);
+        pagePartitioner.partitionPage(page);
+
+        // TODO: PartitionedOutputOperator reports incorrect output data size #11770
```

Comment:
@yingsu00 @ajaygeorge #11770 has been fixed. What would it take to port the fix?

Reply:
OptimizedPartitionedOutputOperator requires different fix because it doesn't use pageBuilders. Also tests need to be updated.